pylops/waveeqprocessing/_kirchhoff_cuda.py

```diff
@@ -0,0 +1,605 @@
+
+        This method handles data preparation and execution of CUDA kernels using streams,
+        for both forward and adjoint operations of dynamic case.
+
+        Parameters
+        ----------
+        x : :obj:`numpy.ndarray`
+            Input data (image or seismic data).
+        opt : :obj:`str`
+            Operation type, either '_matvec' for forward or '_rmatvec' for adjoint.
+
+        Returns
+        -------
+        y : :obj:`numpy.ndarray`
+            Output data after processing.
+
+        """
+
+        if opt == "_matvec":
+            x = x.ravel()
+            y = np.zeros((self.ns * self.nr, self.nt))
+        elif opt == "_rmatvec":
+            y = np.zeros(self.ni)
+
+        y_d_dict = {}
+        isrc_list_d_dict = {}
+        for stream, isrc_list in self.sources_per_streams.items():
+            x_d = cuda.to_device(x, stream=stream)
+            y_d_dict[stream] = cuda.to_device(y, stream=stream)
+            isrc_list_d_dict[stream] = cuda.to_device(isrc_list, stream=stream)
+
+        for stream in self.sources_per_streams.keys():
+            stream.synchronize()
+        for irec in range(self.nr):
+            for stream, isrc_list in self.sources_per_streams.items():
+                if opt == "_matvec":
+                    self._ampsrcrec_kirch_matvec_cuda_streams[
+                        self.num_blocks, self.num_threads_per_blocks, stream
+                    ](
+                        *self.const_inputs,
+                        self.trav_srcs_d_global,
+                        self.amp_srcs_d_global,
+                        self.angles_srcs_d_global,
+                        self.trav_recs_d_global,
+                        self.amp_recs_d_global,
+                        self.angles_recs_d_global,
+                        y_d_dict[stream],
+                        isrc_list_d_dict[stream],
+                        irec,
+                        x_d
+                    )
+                elif opt == "_rmatvec":
+                    self._ampsrcrec_kirch_rmatvec_cuda_streams[
+                        self.num_blocks, self.num_threads_per_blocks, stream
+                    ](
+                        *self.const_inputs,
+                        self.trav_srcs_d_global,
+                        self.amp_srcs_d_global,
+                        self.angles_srcs_d_global,
+                        self.trav_recs_d_global,
+                        self.amp_recs_d_global,
+                        self.angles_recs_d_global,
+                        y_d_dict[stream],
+                        isrc_list_d_dict[stream],
+                        irec,
+                        x_d
+                    )
+        # Synchronize the streams to ensure all operations have been completed
+        for stream in self.sources_per_streams.keys():
+            stream.synchronize()
+        y_streams = []
+        for stream, y_dev in y_d_dict.items():
+            y_streams.append(y_dev.copy_to_host(stream=stream))
```

Comment:
When copying from host, it's better to provide a target array which is pinned. Example:

```python
for stream, y_dev in y_d_dict.items():
    ary = cuda.pinned_array_like(y_dev)
    y_dev.copy_to_host(ary, stream=stream)
    y_streams.append(ary)
```